contracts/utils/Test/Cruzo1155VaultTest.sol

```diff
@@ -0,0 +1,114 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/IERC1155Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/utils/ERC1155HolderUpgradeable.sol";
+
+error Vault__GivenHashIsNotEmpty();
+error Vault__NoGiftByGivenHash();
+error Vault__CallerIsNotMarket();
+
+contract Cruzo1155TempVault_Test is
+    Initializable,
+    ContextUpgradeable,
+    UUPSUpgradeable,
+    ERC1155HolderUpgradeable,
+    OwnableUpgradeable
+{
+    modifier isCallerMarket() {
+        if (_msgSender() == marketAddress) {
+            revert Vault__CallerIsNotMarket();
+        }
+        _;
+    }
+
+    modifier isGiftExists(string calldata secretKey) {
+        if (
+            hashToToken[keccak256(bytes(secretKey))].tokenAddress == address(0)
+        ) {
+            revert Vault__NoGiftByGivenHash();
+        }
+        _;
+    }
+    modifier isGivenHashEmpty(string calldata _hash) {
+        if (hashToToken[bytes32(bytes(_hash))].tokenAddress != address(0)) {
+            revert Vault__GivenHashIsNotEmpty();
+        }
+        _;
+    }
+
+    constructor() {}
+
+    function initialize(address inittialMarketAddress) public initializer {
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+        __Context_init();
+        marketAddress = inittialMarketAddress;
+    }
+
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    event GiftVaulted(
+        address indexed tokenAddress,
+        uint256 indexed tokenId,
+        uint256 indexed amount
+    );
+    event GiftClaimed(
+        address indexed claimer,
+        address indexed tokenAddress,
+        uint256 indexed tokenId,
+        uint256 amount
+    );
+    struct TokenCredentials {
+        uint256 tokenId;
+        uint256 amount;
+        address tokenAddress;
+    }
+
+    address public marketAddress;
+
+    mapping(bytes32 => TokenCredentials) hashToToken;
+
+    function claimGift(string calldata userSecretKey)
```

Comment:
Rename `userSecretKey` -> `_secretKey`

Reply:
Please add an option to pass-in `_targetAddress`. 
If it's passed then we use it to transfer NFT to this address. Fallback (if not passed) is `_msgSender()`.
We might need it to gift to a user who has no cryptocurrency to run transaction on his/her own.

Reply:
A bit confused. Can u explain more properly please?(You mean put this implementation in fallback function and function will be with this optional argument?)

Reply:
I would do it using function overloading, e.g.:

`
function claim(_secretKey) {}
`
`
function claim(_secretKey, _targetAddress) {}
`
and both would call internal implementation 

`
function _claim(_secretKey, _targetAddress) internal {}
`

See https://docs.soliditylang.org/en/v0.4.21/contracts.html#function-overloading

Reply:
I cant do it like u mentioned, need to do meetup to describe this situation